google-cloud-firestore/src/main/java/com/google/cloud/firestore/UpdateBuilder.java

```diff
@@ -248,31 +238,26 @@ private T performSet(
     DocumentTransform documentTransform =
         DocumentTransform.fromFieldPathMap(documentReference, documentData);
 
-    if (options.isMerge()) {
-      if (options.getFieldMask() != null) {
-        List<FieldPath> fieldMask = new ArrayList<>(options.getFieldMask());
-        fieldMask.removeAll(documentTransform.getFields());
-        documentMask = new FieldMask(fieldMask);
-      } else {
-        documentMask = FieldMask.fromObject(fields);
-      }
+    if (options.getFieldMask() != null) {
+      List<FieldPath> fieldMask = new ArrayList<>(options.getFieldMask());
+      fieldMask.removeAll(documentTransform.getFields());
+      documentMask = new FieldMask(fieldMask);
+    } else if (options.isMerge()) {
+      documentMask = FieldMask.fromObject(fields);
     }
 
-    Mutation mutation = addMutation();
-
-    boolean hasDocumentData = !documentSnapshot.isEmpty() || !documentMask.isEmpty();
-
-    if (!options.isMerge()) {
-      mutation.document = documentSnapshot.toPb();
-    } else if (hasDocumentData || documentTransform.isEmpty()) {
-      mutation.document = documentSnapshot.toPb();
-      mutation.document.setUpdateMask(documentMask.toPb());
+    verifyNotCommitted();
```

Comment:
This feels out of place now. Move to top of method?

Reply:
done.